virt-v2v/cold/entrypoint.go

```diff
@@ -0,0 +1,274 @@
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	OVA     = "ova"
+	vSphere = "vSphere"
+	DIR     = "/var/tmp/v2v"
+	FS      = "/mnt/disks/disk[0-9]*"
+	Block   = "/dev/block[0-9]*"
+	VDDK    = "/opt/vmware-vix-disklib-distrib"
+)
+
+type Domain struct {
+	XMLName xml.Name `xml:"domain"`
+	Name    string   `xml:"name"`
+	OS      OS       `xml:"os"`
+}
+
+type OS struct {
+	Type   OSType `xml:"type"`
+	Loader Loader `xml:"loader"`
+	Nvram  Nvram  `xml:"nvram"`
+}
+
+type OSType struct {
+	Arch    string `xml:"arch,attr"`
+	Machine string `xml:"machine,attr"`
+	Content string `xml:",chardata"`
+}
+
+type Loader struct {
+	Readonly string `xml:"readonly,attr"`
+	Type     string `xml:"type,attr"`
+	Secure   string `xml:"secure,attr"`
+	Path     string `xml:",chardata"`
+}
+
+type Nvram struct {
+	Template string `xml:"template,attr"`
+}
+
+var (
+	xmlFilePath      string
+	requestProcessed = make(chan bool)
+)
+
+func main() {
+	virtV2vArgs := []string{"virt-v2v", "-v", "-x"}
+	source := os.Getenv("V2V_source")
+
+	requiredEnvVars := map[string][]string{
+		vSphere: {"V2V_libvirtURL", "V2V_secretKey", "V2V_vmName"},
+		OVA:     {"V2V_diskPath", "V2V_vmName"},
+	}
+
+	if envVars, ok := requiredEnvVars[source]; ok {
+		if !checkEnvVariablesSet(envVars...) {
+			fmt.Printf("Following environment variables need to be defined: %v\n", envVars)
+			os.Exit(1)
+		}
+	}
+
+	switch source {
+	case vSphere:
+		virtV2vArgs = append(virtV2vArgs, "-i", "libvirt", "-ic", os.Getenv("V2V_libvirtURL"))
+		pwFilePath := fmt.Sprintf("%s/vmware.pw", DIR)
+		if err := os.WriteFile(pwFilePath, []byte(os.Getenv("V2V_secretKey")), 0600); err != nil {
+			fmt.Printf("Failed to write to disk: %v\n", err)
+			os.Exit(1)
+		}
+		virtV2vArgs = append(virtV2vArgs, "-ip", pwFilePath)
+
+		if _, err := os.Stat(VDDK); err == nil {
+			virtV2vArgs = append(virtV2vArgs,
+				"-it", "vddk",
+				"-io", fmt.Sprintf("vddk-libdir=%s", VDDK),
+				"-io", fmt.Sprintf("vddk-thumbprint=%s", os.Getenv("V2V_thumbprint")),
+			)
+		}
+		virtV2vArgs = append(virtV2vArgs, "--", os.Getenv("V2V_vmName"))
+	case OVA:
+		virtV2vArgs = append(virtV2vArgs, "-i", "ova", os.Getenv("V2V_diskPath"))
+	}
+
+	fmt.Println("Preparing virt-v2v")
+	if err := os.MkdirAll(DIR, os.ModePerm); err != nil {
+		fmt.Println("Error creating directory  ", err)
+		os.Exit(1)
+	}
+	virtV2vArgs = append(virtV2vArgs, "-o", "local", "-os", DIR)
+
+	//Disks on filesystem storage.
+	if err := LinkDisks(FS, 15); err != nil {
+		os.Exit(1)
+	}
+	//Disks on block storage.
+	if err := LinkDisks(Block, 10); err != nil {
+		os.Exit(1)
+	}
+
+	if err := executeVirtV2v(virtV2vArgs); err != nil {
+		os.Exit(1)
+	}
+
+	if source == OVA {
+		var err error
+		xmlFilePath, err = waitForXMLFile(DIR, "xml", 5, 60)
+		if err != nil {
+			fmt.Println("Error waiting for XML file:", err)
+			os.Exit(1)
+		}
+
+		http.HandleFunc("/firmware", firmwareHandler)
+		server := &http.Server{Addr: ":8080"}
+
+		go func() {
```

Comment:
I'd have the caller invoke the shutdown rather than the pod shut down itself

Reply:
if we go this way, we can also drop the call to get the OVF but take it from the status instead and later call shutdown from the controller..

Reply:
@bennyz recommended adding an endpoint to shut down the server after confirming data receipt from the request, which also allows for retries if necessary. It's a minor adjustment that makes sense to ensure data fetching. We could also change this to your suggestion, but I'm not sure it's worth the effort right now since its involving changing the migration flow, As I  mentioned in other comment, shutting down the pod is required to mark the stage complete but we can always make these improvements later.

Reply:
I've added the change of another endpoint and shutting down from the client side.